### PR TITLE
Support for --jsxFactory option

### LINKED ISSUE #9582: Support Custom JSX Pragma
Hi. I've searched through the various issues (as well as other places this has been discussed) and I don't see a current open issue to address the lack of custom JSX pragma support.

As discussed elsewhere, other transpilers allow specifying any function name to insert when producing hyperscript from JSX.  It's worth noting that hyperscript predates React, and is a format shared by a large number of libraries entirely separate from the React project.  Currently, TypeScript only allows configuration of an Object on which to invoke `.createElement()`, however the [original implementations](https://github.com/dominictarr/hyperscript) and many descendants thereof simply use `h()`.  You can read my attempt to document some of the context and background [here](http://www.jasonformat.com/wtf-is-jsx/).

The configuration would presumably look something like this:

``` js
{
    "compilerOptions": {
        "jsx": "react",
        "jsxPragma": "h"
    }
}
```

... and the output would look like:

``` js
// input:
<div id="foo">hello</div>

// output:
h('div', { id:'foo' }, 'hello');
```

Has this been bypassed for feasibility reasons?  If not, I think it would be a valuable addition.

### PR DESCRIPTION
**Pull request by sheetalkamat**
Fixes #9582  and #12067

### CODE REVIEW

File: src/compiler/types.ts [hunk 1]
```diff
@@ -2250,6 +2251,7 @@ namespace ts {
     /* @internal */
     export interface TypeCheckerHost {
         getCompilerOptions(): CompilerOptions;
+        getJsxFactoryEntity(): EntityName;
```
Comment (mhegazy) [inline src/compiler/types.ts]: this feels odd. i would rather we parsed the jsxFactory string one more time in the checker instead.

**Comment by appsforartists** [comment]
Do we still need typescript@next for this?  https://github.com/Microsoft/TypeScript/wiki/Roadmap says it's landed, but I'm seeing 

```
error TS5023: Unknown compiler option 'jsxFactory'.
```

in 2.1.1.

**Comment by mhegazy** [comment]
The roadmap says it is in `typescript@next`.  the issues (https://github.com/Microsoft/TypeScript/issues/9582 and https://github.com/Microsoft/TypeScript/issues/12067) are marked with milestone `TypeScript 2.1.3` which is not `TypeScript 2.1.1` and has not shipped yet.

**Comment by appsforartists** [comment]
Thanks for the clarification.

I'm not sure the roadmap makes that clear.  The heading "2.1 (November 2016)" reads like those features have already shipped in 2.1.0, especially since they have little checkmarks next to them.  Someone would have to scroll all the way to the bottom to realize the checkmarks are meant to connote "next" features.

Moreover, the PR linked from the Roadmap doesn't have a version label/milestone attached to it.

**Comment by mhegazy** [comment]
At the bottom of the roadmap page, there is an explanation that the checkmark means it is available in `typescript@next`.

**Comment by mhegazy** [comment]
did not mean to sound pedantic. just wanted to make sure the process is clear. questions are always welcomed.

**Comment by appsforartists** [comment]
I saw that, but only after your first response.  It never occurred to me to scroll to the bottom of the page until you said "The roadmap says it is in `typescript@next`".

Have you considered putting the checkbox descriptor at the top?  How about making your roadmap more granular (e.g. change the Nov heading to **2.1.1: stable** and have a heading above it for **next**)?

I don't mean to be nit picky, but I'm sure I'm not the last person who will misread the current Roadmap.

**Comment by mhegazy** [comment]
thanks for the tip. moved the note to the top. and will consider having better naming for milestones in the future.

**Pull request merged**